src/kernels/jupyter/session/jupyterRequestCreator.web.ts

```diff
@@ -0,0 +1,118 @@
+import { IJupyterRequestCreator } from '../types';
+
+const JupyterWebSockets = new Map<string, WebSocketIsomorphic & IKernelSocket>(); // NOSONAR
+
+// Function for creating node Request object that prevents jupyterlab services from writing its own
+// authorization header.
+/* eslint-disable @typescript-eslint/no-explicit-any */
+@injectable()
+export class JupyterRequestCreator implements IJupyterRequestCreator {
+    public getRequestCtor(getAuthHeader?: () => any) {
+        class AuthorizingRequest extends Request {
+            constructor(input: RequestInfo, init?: RequestInit) {
+                super(input, init);
+
+                // Add all of the authorization parts onto the headers.
+                const origHeaders = this.headers;
+                const authorizationHeader = getAuthHeader!();
+                const keys = Object.keys(authorizationHeader);
+                keys.forEach((k) => origHeaders.append(k, authorizationHeader[k].toString()));
+                origHeaders.set('Content-Type', 'application/json');
+
+                // Rewrite the 'append' method for the headers to disallow 'authorization' after this point
+                const origAppend = origHeaders.append.bind(origHeaders);
+                origHeaders.append = (k, v) => {
+                    if (k.toLowerCase() !== 'authorization') {
+                        origAppend(k, v);
+                    }
+                };
+            }
+        }
+
+        return getAuthHeader ? AuthorizingRequest : Request;
+    }
+
+    public getWebsocketCtor(cookieString?: string, allowUnauthorized?: boolean, getAuthHeaders?: () => any) {
+        class JupyterWebSocket extends KernelSocketWrapper(WebSocketIsomorphic) {
+            private kernelId: string | undefined;
+            private timer: NodeJS.Timeout | number;
+
+            constructor(url: string, protocols?: string | string[] | undefined) {
+                let co = {};
+                let co_headers: { [key: string]: string } | undefined;
+
+                if (allowUnauthorized) {
+                    co = { ...co, rejectUnauthorized: false };
+                }
+
+                if (cookieString) {
+                    co_headers = { Cookie: cookieString };
+                }
+
+                // Auth headers have to be refetched every time we create a connection. They may have expired
+                // since the last connection.
+                if (getAuthHeaders) {
+                    const authorizationHeader = getAuthHeaders();
+                    co_headers = co_headers ? { ...co_headers, ...authorizationHeader } : authorizationHeader;
+                }
+                if (co_headers) {
+                    co = { ...co, headers: co_headers };
+                    console.log(`CO Headers: ${co}`);
+                }
+
+                super(url, protocols);
+
+                // TODO: How to send auth headers? Try debugging this to see what happens.
```

Comment:
Old TODO? Looks like you are handling the headers above.

Reply:
No this is still a todo. The constructor for the node websocket takes the CO headers. The constructor for the browser websocket doesn't. 

I _think_ they can be applied to each request during the request constructor, but not sure yet.